WIP: Omni jammer patch

Crowdedlight opened the pull request:
Work in progress on changing to use server authority for jammers and beacons. The server will contain the source of truth, and run the auto cleanups. 

So clients will now ask the server to remove or create jammers, and the server will update and broadcast the updates to all clients. To support JIP, a client joining the game will on postInit ask the server for an update, and the server will respond with a targeted event containing the current state of the jammers. 

This should remove race-conditions and problematic async updating across clients. Make the code cleaner and remove extra checks from the PFH on clients, potentially increasing performance. 

This is talked about and needed for: https://github.com/Crowdedlight/Crows-Electronic-Warfare/pull/65 
To avoid merge conflicts I made this as a patch to apply to your PR. So as long as we merge this first, then the PR should reflect the states we implement here. 

**This is WIP, and have not been tested on dedicated server yet**

Edit: Jammer with new architecture have been tested on local multiplayer game. Missing rewrite of Spectrum beacons before testing on dedicated

todo:

- [x] Change beacons to be server auth like new jammers
- [x] Check behaviour relying on beacons, like c-track etc. and how to handle server-auth setup
- [x] This changes major parts of the networking part, so require dedicated server test for all functions to verify (I will get that going on our server when its in a state ready for test)
- [x] Change ``initDroneSignals`` to use the cba script version for class eventhandlers, and let it be mandated by a CBA option where you can enter string array of classnames. Add most drones, weaponized and darters to the list by default. Then users can dial in their defaults as they want. (CBA setting should be sync'ed, as behaviour is server-wide)
- [x] Change test mission to work with relative path with latest hemtt release. Supports v1.11.2 of hemtt only, but ``hemtt launch`` will now work across machines/developers

Crowdedlight committed:
updated git ignore

Crowdedlight committed:
changed jammers to use server authority as source of truth

Crowdedlight committed:
added missing function for state request

Crowdedlight committed:
removed lingering code in jamming server loop

Crowdedlight committed:
fixed errors during local MP test

Crowdedlight commented on addons/main/functions/fnc_jammerPlayerLocal.sqf:
```diff
@@ -10,35 +10,9 @@ main script that handles the jamming whenever there is at least 1 active jammer
-} forEach _removeList;
-
 //IF ZEUS, DON'T JAM...update markers and skip.
 if (call EFUNC(zeus,isZeus)) exitWith {
```
@b-mayr-1984 

I noticed that you supported that zeus' not in zeus camera, could get jammed when controlling drones. However as it is currently this line means that nothing later in the script will be run. As I initially have Zeus set to never be jammed, even when not in zeus view. 

With the idea of zeus' radio never be jammed so zeus can teleport around as he pleases without thinking about placement. 

If your goal is to also support zeus, not in zeus cam, to be jammed on drones, we can change this so it just skips the TFAR jamming but leaves the rest running?

b-mayr-1984 replied:
@Crowdedlight I wasn't a fan of Zeus not being affected by drone jamming (just made testing more complicated). But I thought I just stick with the original logic. 

So I am fine with changing it to only ignoring TFAR jamming. 🙂

Crowdedlight replied:
Normally I just disable the zeus immunity when debugging/testing ;-) 

But that sounds fair. I'll change it so its only tfar jamming that doesn't affect zeus at any point.

Crowdedlight replied:
done in 6b82275

It should be ready to test on a dedicated server for the jamming part. I still haven't done the architecture change to the spectrum beacons, but planning to do that this weekend. 
Although with the current changes for jamming, the spectrum beacons might be somewhat fixed now too, as only the server would spawn them for jammers, instead of everyone. But should still be changed to the server-authority method.

b-mayr-1984 replied:
> Normally I just disable the zeus immunity when debugging/testing ;-)

Same. 
I replaced `exitWith` with `then` temporarily when testing.

Crowdedlight commented:
So currently its only drones that are "jammable" that will be jammed by the omni-jammers. Which means they need a spectrum signal on them. Zeus or editor can use the modules to set new ones jammable, but I am thinking we might want to increase the drones that has signals by default? 

Its currently set to the following: 
```cpp
class UGV_01_base_F {
	class crowsew_dronesignal_eh {
		init = QUOTE(_this call FUNC(initDroneSignals));
	};
};
class UGV_01_rcws_base_F {
	class crowsew_dronesignal_eh {
		init = QUOTE(_this call FUNC(initDroneSignals));
	};
};
class UAV_03_dynamicLoadout_base_F {
	class crowsew_dronesignal_eh {
		init = QUOTE(_this call FUNC(initDroneSignals));
	};
};	
class UGV_02_Demining_Base_F {
	class crowsew_dronesignal_eh {
		init = QUOTE(_this call FUNC(initDroneSignals));
	};
};	
class UAV_06_antimine_base_F {
	class crowsew_dronesignal_eh {
		init = QUOTE(_this call FUNC(initDroneSignals));
	};
};
```

Which covers most of the weaponized ones, air and ground. Should that be increased to by default include all drones, darters etc? 
Or should it be made into a CBA setting in some way? Maybe an override default behaviour and textbox to input array of classnames you want to default init on?

Crowdedlight committed:
changed so zeus is only immune to tfar jamming, not drone jamming. Zeus in zeus-view will still be immune to drone jamming if remote-control

b-mayr-1984 commented:
> Zeus or editor can use the modules to set new ones jammable, but I am thinking we might want to increase the drones that has signals by default?

> Should that be increased to by default include all drones, darters etc? Or should it be made into a CBA setting in some way? Maybe an override default behaviour and textbox to input array of classnames you want to default init on?

So far I was happy with the ability to add entire classes in 3den using your module.

Making more by default might be what a lot of mod users expect when they first test the mod.

But no matter which route is taken I recommend to always leave the door open to reduce a large default (e.g. if a mission designer considers/wants some drones to be immune to jamming).

The cleanest way is I think to remove any hard coded drones from SQF files (including UGVs).
Then add a large list as a CBA setting with class names (as you described) as default value for this setting.
Less technical users then have a lot of drones to start with and more tech savvy users can shorten/modify the list if they want.

Crowdedlight commented:
> So far I was happy with the ability to add entire classes in 3den using your module.
> 
> Making more by default might be what a lot of mod users expect when they first test the mod.
> 
> But no matter which route is taken I recommend to always leave the door open to reduce a large default (e.g. if a mission designer considers/wants some drones to be immune to jamming).
> 
> The cleanest way is I think to remove any hard coded drones from SQF files (including UGVs). Then add a large list as a CBA setting with class names (as you described) as default value for this setting. Less technical users then have a lot of drones to start with and more tech savvy users can shorten/modify the list if they want.

Yeah, I think that is the more correct way to go. Puts the options in the users hands by default, and zeus/editor can always use the module to add extra specific drones to the list, or remove them on the fly. (just remove the spectrum beacon). 

I does mean I need to change them from config entries, to script entries with CBAs class eventhandlers. but that is fine. Essentially same functionality, just needs to be put in a function that gets run at load from CBA settings. Adding it to the todo-list above.

Crowdedlight committed:
moved default jammable drones to CBA option

Crowdedlight committed:
moved test mission into .hemtt folder for new hemtt launch support with relative path

Crowdedlight commented:
btw, if you got any good ideas how to handle https://github.com/Crowdedlight/Crows-Electronic-Warfare/issues/66 to ensure a minimum separation let me know. 

I had considered sorting all signals by frequency, and then going through them saving the "gaps" between them. Then taking the biggest gap, shorten it with a "minimum_seperation" in both ends and run a random function within it. Should give an almost ~O(n) performance, not counting the sorting. 
But unsure that becomes too much of a pattern as new signals would always appear in the biggest gap. 

```sqf
// randomize frequency 
// TODO randomize within a window that gives minimum seperation to existing signals. 
private _freq = 433.00 + (random 7);
```

b-mayr-1984 commented:
> btw, if you got any good ideas how to handle [Crowdedlight#66](https://github.com/Crowdedlight/Crows-Electronic-Warfare/issues/66) to ensure a minimum separation let me know.

With the fix of #59 this isn't a big concern for me anymore if I am being honest. Since the LMB action is always performed on the strongest signal now there isn't much need for zooming in anyways. Particularly when keeping in mind that players can do spatial filtering by moving the antenna.

The frequency separation would have been more necessary if no other means of selection/separation were possible. But this problem is gone now.

I could live with a `won't fix` here.

Crowdedlight committed:
changed beacons to server authority architecture for better jip handling

Crowdedlight committed:
forgot to undefine debug

Crowdedlight commented:
Redone the beacons to use server authority architecture, which, like the jammers, should solve race conditions, multiple sources of truth and JIP shenanigans. Also makes the communication more clean/clear. 

Any client joining the server will on init now ask the server for current beacons, and then receive all future updates from the server. eliminates the JIP events and removes part of the cleanup that happened on all clients, to now just be running on server, and server broadcast the new state. 

I have tested it on hosted multiplayer, but not dedicated server. I hope to get time the coming week to do a session on dedicated server to validate everything works, but with other irl work deadlines and being away next weekend, I am unsure if I will manage. 

Feel free to review this PR for errors/stuff I might have missed. And let me know if you end up testing some of the points below on a dedicated server before me. Then we can tick them off as we get though them. :)

-------
The following features needs to be tested on dedicated server:

- [x] add/remove beacons as zeus
- [x] add/remove beacons with editor modules
- [x] JIP joining with various states of beacons
- [x] C-Track attached to vehicles and one self
- [x] JIP joining while someone has C-track active on self and/or vehicle
- [x] Add/remove jammer as zeus
- [x] Add/remove jammer as editor module
- [x] JIP joining with various states of jammers added and disabled
- [x] Signal tracking and listening with TFAR tracking (Signals appear/disappear correctly on TFAR transmissions)
- [x] Listening to TFAR still works with ICOM radio
- [x] Signal tracking and listening with RadioChatterModule
- [x] Test Omni Jammers for AI drones and players
- [x] Test Omni Jammers for JIP players
- [ ] Test spectrum device jammer (jams AI behaviour, and dc player controlled drones (Only effecting drones ability to obtain commands, not to send data back to controller)
- [ ] Test CBA option for default jammable classnames

b-mayr-1984 commented:
> * [ ]  Test spectrum device jammer (IIRC should only jam AI behaviour, not player vision, as its directed at the drone, not the controller. (Only effecting drones ability to obtain commands, not to send data back to controller)

Player pilots (both "Driver" and "Gunner") should get disconnected from the drone though. Otherwise a player piloted drones would be immune to jamming.

b-mayr-1984 commented:
I did some testing on a dedicated server regarding these 2 tasks:
> * [ ]  add/remove beacons as zeus
> * [ ]  add/remove beacons with editor modules

### Observation 1
I see dual initializations like in #61 if there is a synched `Apply to Classnames` module present in `mission.sqm`.
![image](https://github.com/Crowdedlight/Crows-Electronic-Warfare/assets/76476468/87cdd612-e9ce-4778-9a6d-a69a9aa8072f)

![image](https://github.com/b-mayr-1984/Crows-Electronic-Warfare/assets/76476468/fd9acb32-0ef9-480e-827e-61a04724df4f)

I assume this is due to the new setting `crowsEW_spectrum_defaultClassForJammingSignal` which works on top of that module and adds another beacon.

### Observation 2
Stompers that where placed in 3den don't have a `beacon` on mission start.
Stompers that have been placed at runtime with Zeus show a `beacon`.
I would have expected that `crowsEW_spectrum_defaultClassForJammingSignal` also applies to the entities in `mission.sqm` and not just those placed via Zeus.

### Observation 3
Using the `Set Spectrum Signal Source` Zeus module creates a red label in 3D space but the signal is not visible in the Spectrum Device (using the jamming antenna). I suspect that most likely the new the signal is not of type `drone` and thus not visible using this antenna. Even though as a developer I understand this I can imagine this is not what the average user would call proper behaviour.


### Conclusion
Other than these 3 observations the rest seems to work just fine. 🙂

Crowdedlight commented:
> I see dual initializations
> I assume this is due to the new setting `crowsEW_spectrum_defaultClassForJammingSignal` which works on top of that module and adds another beacon.

Good point. I'll add to the initSignals function that it doesn't add signals on a drone that already have a signal. Will fix that. 

> ### Observation 2
> Stompers that where placed in 3den don't have a `beacon` on mission start. Stompers that have been placed at runtime with Zeus show a `beacon`. I would have expected that `crowsEW_spectrum_defaultClassForJammingSignal` also applies to the entities in `mission.sqm` and not just those placed via Zeus.

Good catch. I didn't set the "retroactive" option on the script version of the eventhandler. Will fix that. 
https://cbateam.github.io/CBA_A3/docs/files/xeh/fnc_addClassEventHandler-sqf.html
Also need to double-check if that function should be run on all clients, or is enough to have it one place... 

> ### Observation 3
> 
> Using the `Set Spectrum Signal Source` Zeus module creates a red label in 3D space but the signal is not visible in the Spectrum Device (using the jamming antenna). I suspect that most likely the new the signal is not of type `drone` and thus not visible using this antenna. Even though as a developer I understand this I can imagine this is not what the average user would call proper behaviour.

hmm. each antenna has its own "frequency" area, so maybe its outside the frequency for the jammable signals. I believe I should have set it so if its within the frequency the antenna can receive, then you would see it, you just wouldn't be able to jam it. 

But I will double check.

Crowdedlight committed:
added so jammable drones applies retroactively for units put down before we register the function, and making sure the init is only fired on the server, instead of on each client

Crowdedlight committed:
added check to not auto add jammable signal on drone if it already have one

Crowdedlight commented:
Added fixes for obs 1 and 2. I haven't checked 3 yet. If a signal is showing on the device should only be based on frequency that current antenna can equip. Not type. 

I am leaving now for the weekend but hope to be able to test this myself on dedicated server next week. :)

b-mayr-1984 commented:
> > ### Observation 3
> > Using the `Set Spectrum Signal Source` Zeus module creates a red label in 3D space but the signal is not visible in the Spectrum Device (using the jamming antenna). I suspect that most likely the new the signal is not of type `drone` and thus not visible using this antenna. Even though as a developer I understand this I can imagine this is not what the average user would call proper behaviour.
> 
> hmm. each antenna has its own "frequency" area, so maybe its outside the frequency for the jammable signals. I believe I should have set it so if its within the frequency the antenna can receive, then you would see it, you just wouldn't be able to jam it.
> 
> But I will double check.

It appears to be intentional behaviour that it is not displayed. Because also the code comment mentions it explicitly.
https://github.com/Crowdedlight/Crows-Electronic-Warfare/blob/omni_jammer_patch/addons/spectrum/functions/fnc_spectrumTrackingLocal.sqf#L44-L45

Right now players might get confused by not seeing this signal source. 
If the logic is changed players might instead get confused by signals they see in the jammer, but that don't react to jamming (even though this is pretty realistic).

I'd say we ignore Observation 3 for now because it has nothing to do with any PR. Further discussion deserves its own issue IMHO.

b-mayr-1984 commented:
@Crowdedlight please change 
https://github.com/Crowdedlight/Crows-Electronic-Warfare/blob/omni_jammer_patch/addons/main/functions/fnc_jammerPlayerLocal.sqf#L96
so that the debug message is not displayed (or only displayed for Zeuses).

I lack the permission to push that commit myself.

b-mayr-1984 commented:
> Added fixes for obs 1 and 2. 

Fixes are confirmed. I just tested this on a dedicated server. Thx 🙂

b-mayr-1984 commented:
> * [ ]  Test spectrum device jammer (jams AI behaviour, and dc player controlled drones (Only effecting drones ability to obtain commands, not to send data back to controller)

Just tested on a dedicated server. 
- Disconnecting players works
- Disabling AI doesn't work

Crowdedlight committed:
removed debug message in jammerPlayerLocal.sqf

Crowdedlight commented:
Thanks for the update and tests!. I pushed fix for removing the debug message. 

I agree with your point in Obs 3. It's a possible ux problem for users, but not something that should be dealt with in this PR. 

So disabling AI doesn't work. I assume zeus couldn't see them as "JAMMED" text either, but just kept seeing them with the "beacon" text. I'll look more into it.

b-mayr-1984 commented:
> So disabling AI doesn't work. I assume zeus couldn't see them as "JAMMED" text either, but just kept seeing them with the "beacon" text. 

I don't know how it looked from Zeus' perspective, since I did the tests alone on the server (as the player operating the Spectrum Device jammer).

Crowdedlight committed:
fixed locality issue with toggle AI after architecture changes

Crowdedlight commented:
I believe I have fixed the locality issue with the AI jamming. I forgot that the new setup had "initDroneSignals" only being run on server, while the toggleAI has to be present on all clients. 
I also realised I had some repeats of eventhandlers that got added once per drone per client, when it should just be added once per client. 

I tested it locally, not on Dedicated yet. I do however get a bit of odd behaviour for flying drones. Setting a quadcopter to fly waypoints and then jamming it with the spectrum device, makes it show as jammed for zeus, but continue to fly the waypoints. 

Doing the same for a UGV, makes the UGV stop where it is, and stop the turret from engaging. Might be edge case with flying drones doing something. But need to test this with more drones and see if I can figure out why flying drones don't stop when jammed, and if it effect all flying drones.

b-mayr-1984 commented:
> I do however get a bit of odd behaviour for flying drones. Setting a quadcopter to fly waypoints and then jamming it with the spectrum device, makes it show as jammed for zeus, but continue to fly the waypoints.
> 
> Doing the same for a UGV, makes the UGV stop where it is, and stop the turret from engaging. Might be edge case with flying drones doing something. But need to test this with more drones and see if I can figure out why flying drones don't stop when jammed, and if it effect all flying drones.

This sounds familiar.
I noticed that the Darter is a bit quirky. 
When the next waypoint is relatively close to the drone when being jammed, it tends to still go there and then land somewhat close to it.
If the next waypoint is hundreds of meters away it does not try to get there and lands earlier. 

I guess this behavior is what you encounter. I consider it "normal" for the Darter.
In case you want to cross check with another type of AI, I recommend a jammable Little Bird. This one reacts more directly to both jamming and release of jamming.

Crowdedlight commented:
> This sounds familiar. I noticed that the Darter is a bit quirky. When the next waypoint is relatively close to the drone when being jammed, it tends to still go there and then land somewhat close to it. If the next waypoint is hundreds of meters away it does not try to get there and lands earlier.
> 
> I guess this behavior is what you encounter. I consider it "normal" for the Darter. In case you want to cross check with another type of AI, I recommend a jammable Little Bird. This one reacts more directly to both jamming and release of jamming.

Cheers. Will try out little bird and some other drones just to check it out. Sounds plausible that this is what I experienced.

b-mayr-1984 commented:
> > * [ ]   Test spectrum device jammer (jams AI behaviour, and dc player controlled drones (Only effecting drones ability to obtain commands, not to send data back to controller)
> 
> Just tested on a dedicated server.
> 
> * Disconnecting players works
> * Disabling AI doesn't work

Tested again with changes from d78ece1aa31c8993ce2f57a07e77432ced004864.
Either feature works now on dedicated server. Nice job 🙂👍

b-mayr-1984 commented:
> * [ ]  Test Omni Jammers for AI drones and players

Just tested this on a dedicated server:
- Film grain noise effect for player piloted drone works
- When entering the kill radius the disconnect `hint` is displayed
  - **But player is not disconnected from drone. Player can continue to control the drone (and film grain is gone which makes flying in the kill zone even easier than outside of it)**
- Once player manually disconnects from the drone the now autonomous drone is successfully jammed and lands
  - Then also reconnect attempts by players fail because of immediate subsequent disconnect.

b-mayr-1984 commented:
> * [ ]  Add/remove jammer as zeus
> * [ ]  Add/remove jammer as editor module

Just tested this on a dedicated server:
- adding works
- removing doesn't

☝️ for both Zeus and 3den modules

Further observations:
- Deleting a jammer object in Zeus mode will not remove its entry in `crowsew_spectrum_beacons` (at least not at client side).
- Jammers defined in 3den show `"any"` as their owning object once the jammer is turned on (see also [#dev-arma3](https://discord.com/channels/976165959041679380/976224730422063214/1218565905488019476))

b-mayr-1984 commented:
> * Jammers defined in 3den show `"any"` as their owning object once the jammer is turned on (see also [#dev-arma3](https://discord.com/channels/976165959041679380/976224730422063214/1218565905488019476))

@dedmen was so kind to help with this issue. 
It appears to be an undefined variable in https://github.com/Crowdedlight/Crows-Electronic-Warfare/blob/omni_jammer_patch/addons/main/functions/fnc_toggleJammerServer.sqf#L30.

He also suggested some more changes in [#dev-arma3](https://discord.com/channels/976165959041679380/976224730422063214/1218572611727003718).

b-mayr-1984 commented:
Further info on this:
> > * [ ]   Add/remove jammer as zeus
> > * [ ]   Add/remove jammer as editor module
> 
> Just tested this on a dedicated server:
> 
> * adding works
> * removing doesn't
> 
> ☝️ for both Zeus and 3den modules
> 
> Further observations:
> 
> * Deleting a jammer object in Zeus mode will not remove its entry in `crowsew_spectrum_beacons` (at least not at client side).
> * Jammers defined in 3den show `"any"` as their owning object once the jammer is turned on (see also [#dev-arma3](https://discord.com/channels/976165959041679380/976224730422063214/1218565905488019476))


Server's RPT shows corresponding error messages:
```
15:42:07 Setting invalid pitch 0.0000 for L Charlie 2-4:1 REMOTE
15:42:07 Error in expression <""], ["_broadcast", true]];

if (_netid == "" || !isServer) exitWith {};


priva>
15:42:07   Error position: <== "" || !isServer) exitWith {};



priva>
15:42:07   Error Generic error in expression
15:42:07 File z\crowsEW\addons\main\functions\fnc_removeJammerServer.sqf..., line 15
```

The only thing that IMHO could go wrong in this LOC is  `_netid == ""` in a case where `_netid` is not of type string. Maybe there is a connection to the previously mentioned undefined variable that showed up in the debugger as `"any"`. 🤔

b-mayr-1984 committed:
prevent nil objects

see discussion at
https://discord.com/channels/976165959041679380/976224730422063214/1218565905488019476

b-mayr-1984 commented:
Fixed in cf3926a65969b64e96faeafb24f61ac4167934e7 (verified with test on dedicated server)
> * Jammers defined in 3den show `"any"` as their owning object once the jammer is turned on (see also [#dev-arma3](https://discord.com/channels/976165959041679380/976224730422063214/1218565905488019476))

b-mayr-1984 committed:
fixed removal of jammers

b-mayr-1984 commented:
Fixed in b807a4b0465fee129c4f7d9c2a1cfe5956078aaa

> Further info on this:
> 
> > > * [ ]    Add/remove jammer as zeus
> > > * [ ]    Add/remove jammer as editor module
> > 
> > 
> > Just tested this on a dedicated server:
> > 
> > * adding works
> > * removing doesn't
> > 
> > ☝️ for both Zeus and 3den modules
> > Further observations:
> > 
> > * Deleting a jammer object in Zeus mode will not remove its entry in `crowsew_spectrum_beacons` (at least not at client side).
> > * Jammers defined in 3den show `"any"` as their owning object once the jammer is turned on (see also [#dev-arma3](https://discord.com/channels/976165959041679380/976224730422063214/1218565905488019476))
> 
> Server's RPT shows corresponding error messages:
> 
> ```
> 15:42:07 Setting invalid pitch 0.0000 for L Charlie 2-4:1 REMOTE
> 15:42:07 Error in expression <""], ["_broadcast", true]];
> 
> if (_netid == "" || !isServer) exitWith {};
> 
> 
> priva>
> 15:42:07   Error position: <== "" || !isServer) exitWith {};
> 
> 
> 
> priva>
> 15:42:07   Error Generic error in expression
> 15:42:07 File z\crowsEW\addons\main\functions\fnc_removeJammerServer.sqf..., line 15
> ```
> 
> The only thing that IMHO could go wrong in this LOC is `_netid == ""` in a case where `_netid` is not of type string. Maybe there is a connection to the previously mentioned undefined variable that showed up in the debugger as `"any"`. 🤔

b-mayr-1984 commented:
> * [ ]  Signal tracking and listening with TFAR tracking (Signals appear/disappear correctly on TFAR transmissions)

Tested on dedicated server.

On the server this error is shown in the RPT log. 
```
20:02:36 Error in expression < param [2,true,[true]];


sleep 1;

if (crowsEW_zeus_hasTFAR == false) exitWith >
20:02:36   Error position: <crowsEW_zeus_hasTFAR == false) exitWith >
20:02:36   Error Undefined variable in expression: crowsew_zeus_hastfar
20:02:36 File /z/crowsEW/addons/editormodules/functions/fnc_enableTrackingTfar.sqf..., line 248
```
This error is not shown at the client side RPT log.

Other functionality I haven't tested because I was testing just by myself and
https://github.com/Crowdedlight/Crows-Electronic-Warfare/blob/omni_jammer_patch/addons/spectrum/functions/fnc_spectrumTrackingLocal.sqf#L36-L37 
prevents self-tracking.
@Crowdedlight do you see a problem with self-tracking? It would definitely be realistic to see ones own emissions as very strong signals.

Crowdedlight committed:
rolled both removals into same if-sentence to avoid idential if-checks right after eachother

Crowdedlight commented:
> On the server this error is shown in the RPT log.

Fixed, Was a check to avoid running zeus functions on clients with no-interface that stopped no-interface clients from setting the variable on init. 

> do you see a problem with self-tracking? It would definitely be realistic to see ones own emissions as very strong signals. 

I chose to remove it as the UI in Arma handles many signals poorly. Having a signal that would always be high in a set frequency would mean that you won't be able to track any other signals in same frequency. Coupled with the fact I didn't see any use cases where you would want to track a signal put on yourself, I disabled it. Was a bit of the thought that your tracker/operator would be good enough to recognize his own signal, and hide it/filter it out for his own tracker. But ofc. his friend next to him would still see it. 

If its important for your usecase I can make it into a CBA setting with default being off. So you have the choice. I personally just didn't see a situation where it would do anything but cause more confusion for players. 

I had thoughts if I wanted to at some point to implement a way to hide or label "known" signals, so an operator could filter out signals not important or at known frequencies. But never had time to do anything with it. Especially as the UI from basegame doesn't look particular fun to edit. 😅 

> When entering the kill radius the disconnect hint is displayed
> But player is not disconnected from drone. Player can continue to control the drone (and film grain is gone which makes flying in the kill zone even easier than outside of it)
This is quite odd. As if the hint is displayed then the line before the hint, would be this, which should disconnect a player. 
```sqf
player connectTerminalToUAV objNull;
```
Did you see anything in rpt that could hint on the problem? This is all done on the client, so locality should not be a problem. Only thought with locality could be a desync between jamMap from server to client. But then the client shouldn't see the jammer at all and not get the grain effect either, nor its removal when inside the radius. 

> Deleting a jammer object in Zeus mode will not remove its entry in crowsew_spectrum_beacons (at least not at client side).

I believe you also fixed that with b807a4b as the server now correctly also remove beacons attached to a jammer.

Crowdedlight committed:
fixed error with mod-loaded params not being set for server in init

Crowdedlight committed:
added null guard in tfar tangent event callback

Crowdedlight commented:
Also thanks to Dedman helping you find that error and the non-guard against null unit. I think my assumption was that the call from TFAR would never pass a null-unit, as then the call wouldn't be fired. But it is good practice to guard in any case. So added that.

b-mayr-1984 commented:
### Self-Tracking
> > do you see a problem with self-tracking? It would definitely be realistic to see ones own emissions as very strong signals.
> 
> I chose to remove it as the UI in Arma handles many signals poorly. Having a signal that would always be high in a set frequency would mean that you won't be able to track any other signals in same frequency. 

Own TFAR emissions wouldn't always be high I hope, but only show when I use my own radio. Am I mistaking here?

That own emissions can make you blind for the emissions of others is correct, yes. I would consider this as intentional behaviour. 
- Just as in PvP ones own talking can prevent you from hearing the enemy. 
- Or ones own engine sounds prevents you from hearing enemy vehicles.

> Coupled with the fact I didn't see any use cases where you would want to track a signal put on yourself, I disabled it. Was a bit of the thought that your tracker/operator would be good enough to recognize his own signal, and hide it/filter it out for his own tracker. [..]

Yeah our sales guys at work would love if that is possible, but it just ain't in reality. 😅 
It is not really an intentional use-case but a maybe unwanted effect of physics.

There are other examples with self-emissions in Arma:
- When I play tank commander or gunner and use thermal imaging I still see my own tank brightly lit in low zoom levels, which makes other thermal contacts stand out less. Do I want that desensitation by my own emissions? Probably not, but that's just what the simulator does (and I enjoy the simulation depth of Arma).
- When playing infantry the muzzle flash from my own fire might disturb my view. It would be more convenient if I only saw other peoples muzzle flash, but this own emission I simply have to deal with.

> If its important for your usecase I can make it into a CBA setting with default being off. So you have the choice. I personally just didn't see a situation where it would do anything but cause more confusion for players.

I guess the question boils down to _**how much simulation depth**_ vs. _**player convenience and arcade play style**_. These are quite frequently stated different points of view. So I guess making it configurable is the best course of action because players then can decide for themselves how they want to play the game. 🙂 

### Flying in kill radius 
> > When entering the kill radius the disconnect hint is displayed
> > But player is not disconnected from drone. Player can continue to control the drone (and film grain is gone which makes flying in the kill zone even easier than outside of it)
> > This is quite odd. As if the hint is displayed then the line before the hint, would be this, which should disconnect a player.
> 
> ```sqf
> player connectTerminalToUAV objNull;
> ```
> 
> Did you see anything in rpt that could hint on the problem? This is all done on the client, so locality should not be a problem. Only thought with locality could be a desync between jamMap from server to client. But then the client shouldn't see the jammer at all and not get the grain effect either, nor its removal when inside the radius.

I need to test again. But good call that it is handled purely local (I forgot about that). Makes it easy to just attach the VSC debugger to it and see what goes wrong.

### Removing beacon from jammer
> > Deleting a jammer object in Zeus mode will not remove its entry in crowsew_spectrum_beacons (at least not at client side).
> 
> I believe you also fixed that with [b807a4b](https://github.com/b-mayr-1984/Crows-Electronic-Warfare/commit/b807a4b0465fee129c4f7d9c2a1cfe5956078aaa) as the server now correctly also remove beacons attached to a jammer.

Yes, correct.

Crowdedlight commented:
> Own TFAR emissions wouldn't always be high I hope, but only show when I use my own radio. Am I mistaking here?

Ah!, I was thinking about putting a zeus-beacon on a player, as that was when I added it initially. I can see the difference for radio signals/tfar usecase. Could add check so it would still show own signal for tfar types, but the constant "zeus" beacons getting ignored?  

> Yeah our sales guys at work would love if that is possible, but it just ain't in reality. 😅 It is not really an intentional use-case but a maybe unwanted effect of physics.

haha yeah, I know. As stated above, I was thinking for the constant beacons, and not the radio ones when talking about it. And while it wouldn't be correct with irl behaviour, I think it's important to also make it work within the constraints of the game.
 
> I guess the question boils down to _**how much simulation depth**_ vs. _**player convenience and arcade play style**_. These are quite frequently stated different points of view. So I guess making it configurable is the best course of action because players then can decide for themselves how they want to play the game. 🙂

Jup, a difficult balance to tread without making everything configurable 😅 
What do you think would make most sense? 
1. hardcode it to only show for tfar/radio transmissions, maybe with a CBA setting to show for all types? 
2. Make it a setting, either you don't see any of your own transmissions, or you see all of them? 

Not quite sure what would be best. 2. is more arcade vs. simulation, but if zeus puts a signal on you constantly you will always see that being very high in the frequency. 

> I need to test again. But good call that it is handled purely local (I forgot about that). Makes it easy to just attach the VSC debugger to it and see what goes wrong.

Good shout. I tested it in hosted multiplayer back then and I seem to recall I got disconnected as player. Not sure what would make that different on dedicated, when it's local. Besides if there is a mismatch between what jammers the client and server see. I'll see if I can reproduce it on my dedicated server too. 

> Yes, correct.

Great! 👍

b-mayr-1984 commented:
> [..] I was thinking about putting a zeus-beacon on a player, as that was when I added it initially. [..] Could add check so it would still show own signal for tfar types, but the constant "zeus" beacons getting ignored?

I needed a while to understand what you mean by a _Zeus beacon_. I suppose you just mean the Zeus module that adds a signal source to an object?!

If that is the case I as a player would still prefer to see my own emissions, because if I am brightly lit this probably means something to the mission/game and I rather know about that than being left in the dark. Seeing it would give me the means to be aware of a maybe cursed (better covert) item that I picked up and that might give my position away to the enemy.

To me the situation is similar to playing a night OP with NVG and IR lasers, where I know the enemy has NVG as well. In this case I need to be aware of my own IR emission to prevent enemies to be alerted to my presence. 
I wouldn't like it to not see my own emission, because then I don't have the means for self-emission-control.

If one really wants to add a signal source that the player, it is attached to, does not see, then I think Zeus can always pick a frequency that is outside of the range that the player can monitor with the antennas they have in their loadout. 

So both _cursed items_ and _players shouldn't know_ kinds of missions are supported then. :slightly_smiling_face: 


> > I guess the question boils down to _**how much simulation depth**_ vs. _**player convenience and arcade play style**_. These are quite frequently stated different points of view. So I guess making it configurable is the best course of action because players then can decide for themselves how they want to play the
 game. 🙂
> 
> Jup, a difficult balance to tread without making everything configurable 😅 What do you think would make most sense?
> 
> 1. hardcode it to only show for tfar/radio transmissions, maybe with a CBA setting to show for all types?
> 2. Make it a setting, either you don't see any of your own transmissions, or you see all of them?

Hm seems to have somewhat of an overlap with the already existing setting `crowsEW_spectrum_tfarSideTrack` (just to keep in mind).

I honestly think emissions should always be visible to anyone with the sensor capable of detecting them. So personally I don't need any filtering/convenience settings. 
But I must note here that I fall pretty much on the simulationist end of the spectrum and I can't represent the arcade perspective very well. I think it's best to reach out so someone that is not so professionally involved into the subject matter like I am and ask for their opinion. :shrug:

Crowdedlight commented:
> I needed a while to understand what you mean by a _Zeus beacon_. I suppose you just mean the Zeus module that adds a signal source to an object?!

Yes, a zeus or script putting another type of beacon on a player that will be on constantly. 
 
> If that is the case I as a player would still prefer to see my own emissions, because if I am brightly lit this probably means something to the mission/game and I rather know about that than being left in the dark. Seeing it would give me the means to be aware of a maybe cursed (better covert) item that I picked up and that might give my position away to the enemy.
...
> So both _cursed items_ and _players shouldn't know_ kinds of missions are supported then. 🙂
...
> I honestly think emissions should always be visible to anyone with the sensor capable of detecting them. So personally I don't need any filtering/convenience settings. But I must note here that I fall pretty much on the simulationist end of the spectrum and I can't represent the arcade perspective very well. I think it's best to reach out so someone that is not so professionally involved into the subject matter like I am and ask for their opinion. 🤷

I can see where you are coming from with a much higher focus on simulation than "arcade". I am personally probably a bit more in the middle. While I do like some simulation, I also don't want the game to turn into something as "cumbersome" (lack of better word) as irl/work, as then its not something the average player would use in a hurry as a fun element. And it can easier become confusing or less intuitive for players without the domain knowledge of RF. I could easily see more casual players reporting it as a bug or be frustrated, if they had a signal on them, and it meant they had a constant high signal that hid other jammable signals in that frequency. 

But I also get the interest of being able to simulate and let players discover if they are being tracked or are broadcasting RF. 

I think the best compromise is to make a CBA setting that enables/disables being able to see your own signal sources. Then you can use that and the "tfar_sidetrack" setting to determine how far you want to go simulation versus making it a more casual game element. As one setting decides if you can see your own signals, and the other if you can see friendly radio signals.

b-mayr-1984 commented:
> I think the best compromise is to make a CBA setting that enables/disables being able to see your own signal sources. Then you can use that and the "tfar_sidetrack" setting to determine how far you want to go simulation versus making it a more casual game element. As one setting decides if you can see your own signals, and the other if you can see friendly radio signals.

Sounds good 🙂👌

b-mayr-1984 commented:
> ### Flying in kill radius
> > > When entering the kill radius the disconnect hint is displayed
> > > But player is not disconnected from drone. Player can continue to control the drone (and film grain is gone which makes flying in the kill zone even easier than outside of it)
> > > This is quite odd. As if the hint is displayed then the line before the hint, would be this, which should disconnect a player.
> > 
> > 
> > ```sqf
> > player connectTerminalToUAV objNull;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Did you see anything in rpt that could hint on the problem? This is all done on the client, so locality should not be a problem. Only thought with locality could be a desync between jamMap from server to client. But then the client shouldn't see the jammer at all and not get the grain effect either, nor its removal when inside the radius.
> 
> I need to test again. But good call that it is handled purely local (I forgot about that). Makes it easy to just attach the VSC debugger to it and see what goes wrong.

I can't reproduce this problem anymore. Maybe the commits in the meantime made the difference. 🤷 
Let's just consider it done for the time being.

Crowdedlight commented:
Sounds good. I am finally on vacation and have some time. So I have set off tomorrow to get the last bits tested on our dedicated server.

b-mayr-1984 commented:
> * [x]  Add/remove jammer as zeus

I found one edge case where removing a jammer leads to an inconsistency.

When an omni-jammer has a drone jammed and than has it's jammer removed via the Zeus module `Remove Jammer` 
- the [spawned jamming function](https://github.com/Crowdedlight/Crows-Electronic-Warfare/blob/omni_jammer_patch/addons/spectrum/functions/fnc_toggleJammingOnUnit.sqf#L45) keeps running and disconnecting players. 
- Also the drone stays in the list `crowsew_spectrum_activeJammedUnits` and is thus still marked as `JAMMED` in the Zeus 3D interface.

b-mayr-1984 commented:
> * [x]  Signal tracking and listening with RadioChatterModule

Appears to work just fine on a dedicated server. 🙂 
Adding and removing via Zeus module (or unit death or deletion) tested successfully.

3den module however I did **_not_** test.

b-mayr-1984 commented:
Just (re-)tested all of these on a dedicated server with another player and myself. They all look good. 🙂 

> * [x]  JIP joining with various states of beacons
> * [x]  C-Track attached to vehicles and one self
> * [x]  JIP joining while someone has C-track active on self and/or vehicle
> * [x]  JIP joining with various states of jammers added and disabled
> * [x]  Signal tracking and listening with TFAR tracking (Signals appear/disappear correctly on TFAR transmissions)
> * [x]  Listening to TFAR still works with ICOM radio
> * [x]  Signal tracking and listening with RadioChatterModule
> * [x]  Test Omni Jammers for AI drones and players
> * [x]  Test Omni Jammers for JIP players


During retesting with more than one player the disconnect of player pilots did not work. 🙁 
Only jamming the drone one is piloting themselves works.
See also [discussion in #dev-arma3](https://discord.com/channels/976165959041679380/976224730422063214/1221109359111573535). 

> * [ ]  Test spectrum device jammer (jams AI behaviour, and dc player controlled drones (Only effecting drones ability to obtain commands, not to send data back to controller)

b-mayr-1984 committed:
prevent "Performance warning" messages in RPT

see https://community.bistudio.com/wiki/Structured_Text for details

Crowdedlight commented:
Note for self, missing tests and tasks to look at today:

- [x] Add setting for self-tracking (Make in another PR when this is done and merged into main repo.) Tracked in open issue
- [x] Investigate disconnecting people from drones issue on dedicated server
- [x] Removing Omni-jammer through zeus module will keep the spawned jamming function running and drone stays in list ``crowsew_spectrum_activeJammedUnits`` and shows "JAMMED" for zeus
- [x] Test spectrum device jammer (jams AI behaviour, and dc player controlled drones (Only effecting drones ability to obtain commands, not to send data back to controller)
- [x] Test CBA option for default jammable classnames
- [x] Signal tracking and listening with RadioChatterModule EDEN module

Crowdedlight committed:
fixed so removing jammers through script/zeus module will stop jamming any drones it was actively jamming

Crowdedlight commented:
Investigation into ``player connectTerminalToUAV [objNull];`` behaviour show the following:

If you run it non-local it will return ``true`` but the player will not be kicked out of the drone view. However, if the player disconnects from the drone view, then the UAV-Terminal shows as not connected.

If the player stays in the drone view then subsequent ``player connectTerminalToUAV [objNull];`` will return false, because the player is already disconnected, they are just not kicked out of the remote cam. 

Should report this as a bug, or at least get them to change so the command is labelled as LE instead of GE. 
The solution for us is to just make a ``targetEvent`` that does the disconnect, so it happens on the players own side. 

**Another Observation:**
~~Seems like it triggers multiple hint on disconnecting player from drone in the time it takes for the player to get out of the view. Will need to check if we can limit that to one. Its not critical, its just hearing the hint sound 4-5 times before it stops.~~

Edit: Think it was the stream I was watching it through. The loop have 0.5 sleep, so can't spam like that.

Crowdedlight committed:
moved player disconnect of drone to local execution

Crowdedlight commented:
@b-mayr-1984 This should be ready for merge now. 

Tested on dedicated server with the change to disconnecting players locally, and that works.
